sdks/java/io/contextual-text-io/src/main/java/org/apache/beam/sdk/io/contextualtextio/RecordWithMetadata.java

```diff
@@ -17,69 +17,103 @@
-@DefaultSchema(AutoValueSchema.class)
-@AutoValue
-public abstract class RecordWithMetadata {
-  public abstract long getRecordOffset();
-
-  public abstract long getRecordNum();
-
-  public abstract String getValue();
-
-  public abstract long getRangeOffset();
-
-  public abstract long getRecordNumInOffset();
-
-  public abstract Builder toBuilder();
-
-  public abstract String getFileName();
-
-  public static Builder newBuilder() {
-    return new AutoValue_RecordWithMetadata.Builder();
+public class RecordWithMetadata {
+
+  public static final String RECORD_OFFSET = "recordOffset";
+  public static final String RECORD_NUM = "recordNum";
+  public static final String VALUE = "value";
+  public static final String RANGE_OFFSET = "rangeOffSet";
+  public static final String RECORD_NUM_IN_OFFSET = "recordNumInOffset";
+  public static final String RESOURCE_ID = "resourceID";
+
+  public static Schema getSchema() {
+    return Schema.builder()
+        .addInt64Field(RECORD_OFFSET)
+        .addInt64Field(RECORD_NUM)
+        .addStringField(VALUE)
+        .addInt64Field(RANGE_OFFSET)
+        .addInt64Field(RECORD_NUM_IN_OFFSET)
+        .addLogicalTypeField(RESOURCE_ID, new ResourceIdRow())
+        .build();
   }
 
-  @AutoValue.Builder
-  public abstract static class Builder {
-    public abstract Builder setRecordNum(long lineNum);
-
-    public abstract Builder setRecordOffset(long recordOffset);
-
-    public abstract Builder setValue(String Value);
-
-    public abstract Builder setFileName(String fileName);
-
-    public abstract Builder setRecordNumInOffset(long recordNumInOffset);
-
-    public abstract Builder setRangeOffset(long startingOffset);
-
-    public abstract RecordWithMetadata build();
+  /** A Logical type using Row to represent the ResourceId type. */
+  public static class ResourceIdRow implements LogicalType<ResourceId, Row> {
+
+    @Override
+    public Schema.FieldType getArgumentType() {
+      return Schema.FieldType.STRING;
+    }
+
+    @Override
+    public String getArgument() {
+      return "";
+    }
+
+    // The underlying schema used to represent rows.
+    private final Schema schema =
+        Schema.builder()
+            .addStringField("schema")
+            .addStringField("fileName")
+            .addBooleanField("isDirectory")
+            .build();
+
+    @Override
+    public String getIdentifier() {
+      return "beam:logical_type:resourceIdRow:v1";
+    }
+
+    @Override
+    public FieldType getBaseType() {
+      return FieldType.row(schema);
+    }
+
+    // Convert the representation type to the underlying Row type. Called by Beam when necessary.
+    @Override
+    public Row toBaseType(org.apache.beam.sdk.io.fs.ResourceId resourceId) {
+      return Row.withSchema(schema)
+          .withFieldValue("schema", resourceId.getScheme())
+          .withFieldValue("fileName", resourceId.getFilename())
+          .withFieldValue("isDirectory", resourceId.isDirectory())
```

Comment:
```suggestion
          .withFieldValue("resource", resourceId.toString())
          .withFieldValue("is_directory", resourceId.isDirectory())
```

Reply:
This looks like something that would be better handled by a `SchemaProvider`. Is there a reason it needs to be a logical type?

Reply:
I think Reza ran into issues with embedding ResourceId as a type within the Row/RecordWithMetadata.

Also, ResourceId is a prime candidate for xlang and being embedded in other schemas. Is SchemaProvider the best way to do this?

Reply:
Its possible Reza was running into BEAM-10765

The main advantage of `SchemaProvider` is it generates the boilerplate for conversions to row and back. If we definitely want this to be a logical type (which sounds reasonable), then it may be possible to replace the `toBaseType` and `toInputType` methods with functions generated by a `SchemaProvider`. That could be left for later though.

Reply:
boilerplate in getSchema() could be removed too.

Reply:
@TheNeuralBit yea, its the Class<T> with field of of T issue.